pkg/services/folder/folderimpl/folder.go

```diff
@@ -334,15 +337,48 @@ func (s *Service) GetChildren(ctx context.Context, q *folder.GetChildrenQuery) (
 		metrics.MFolderIDsServiceCount.WithLabelValues(metrics.Folder).Inc()
 		// nolint:staticcheck
 		f.ID = dashFolder.ID
+
+		hasAccess := parentHasAccess
+		if !hasAccess {
+			g, err := guardian.NewByUID(ctx, f.UID, f.OrgID, q.SignedInUser)
+			if err != nil {
+				return nil, err
+			}
+
+			guardianFunc := g.CanView
+			if q.Permission == dashboardaccess.PERMISSION_EDIT {
+				guardianFunc = g.CanEdit
+			}
+
+			hasAccess, err = guardianFunc()
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if hasAccess {
+			filtered = append(filtered, f)
+		}
 	}
 
-	return children, nil
+	return filtered, nil
 }
 
 func (s *Service) getRootFolders(ctx context.Context, q *folder.GetChildrenQuery) ([]*folder.Folder, error) {
 	permissions := q.SignedInUser.GetPermissions()
-	folderPermissions := permissions[dashboards.ActionFoldersRead]
-	folderPermissions = append(folderPermissions, permissions[dashboards.ActionDashboardsRead]...)
+	var folderPermissions []string
+	if q.Permission == dashboardaccess.PERMISSION_EDIT {
+		folderPermissions = permissions[dashboards.ActionFoldersWrite]
+		folderPermissions = append(folderPermissions, permissions[dashboards.ActionDashboardsWrite]...)
+	} else {
+		folderPermissions = permissions[dashboards.ActionFoldersRead]
+		folderPermissions = append(folderPermissions, permissions[dashboards.ActionDashboardsRead]...)
+	}
+
+	if len(folderPermissions) == 0 && !q.SignedInUser.GetIsGrafanaAdmin() {
+		return nil, nil
+	}
```

Comment:
Question: what is this check for? Can `folderPermissions` ever be an empty list?

Reply:
I don't know if it's anyway possible therefore this is for extra safety in respect with #81304